src/hdmf/term_set.py

```diff
@@ -169,3 +171,103 @@ def __enum_expander(self):
         expander.expand_in_place(self.term_schema_path, enum, output_path)
 
         return output_path
+
+class TermSetWrapper:
+    """
+    This class allows any HDF5 dataset or attribute to have a TermSet.
+    """
+    @docval({'name': 'termset',
+             'type': TermSet,
+             'doc': 'The TermSet to be used.'},
+            {'name': 'value',
+             'type': (list, np.ndarray, dict, str, tuple),
+             'doc': 'The target item that is wrapped, either data or attribute.'},
+            )
+    def __init__(self, **kwargs):
+        self.__value = kwargs['value']
+        self.__termset = kwargs['termset']
+        self.__validate()
+
+    def __validate(self):
+        # check if list, tuple, array
+        if isinstance(self.__value, (list, np.ndarray, tuple)): # TODO: Future ticket on DataIO support
+            values = self.__value
+        # create list if none of those -> mostly for attributes
+        else:
+            values = [self.__value]
+        # iteratively validate
+        bad_values = []
+        for term in values:
+            validation = self.__termset.validate(term=term)
+            if not validation:
+                bad_values.append(term)
+        if len(bad_values)!=0:
+            msg = ('"%s" is not in the term set.' % ', '.join([str(value) for value in bad_values]))
+            raise ValueError(msg)
+
+    @property
+    def value(self):
+        return self.__value
+
+    @property
+    def termset(self):
+        return self.__termset
+
+    @property
+    def dtype(self):
+        return self.__getattr__('dtype')
+
+    def __getattr__(self, val):
+        """
+        This method is to get attributes that are not defined in init.
+        This is when dealing with data and numpy arrays.
+        """
+        return getattr(self.__value, val)
+
+    def __getitem__(self, val):
+        """
+        This is used when we want to index items.
+        """
+        return self.__value[val]
+
+    def __next__(self):
+        """
+        We want to make sure all iterators are still valid.
+        """
+        return self.__value.__next__()
+
```

Comment:
```suggestion
    # uncomment when DataChunkIterator objects can be wrapped by TermSet
    # def __next__(self):
    #     """
    #     We want to make sure all iterators are still valid.
    #     """
    #     return self.__value.__next__()
    #
```
TermSetWrapper does not currently support DataChunkIterators or DataIOs to be wrapped. Let's comment this out for now.